google/cloud/storage/blob.py

```diff
@@ -3092,22 +3093,23 @@ def compose(
         if self.user_project is not None:
             query_params["userProject"] = self.user_project
 
+        _add_generation_match_parameters(
+            query_params,
+            if_generation_match=if_generation_match,
+            if_metageneration_match=if_metageneration_match,
+        )
+
         source_objects = []
         for index, source in enumerate(sources):
             source_object = {"name": source.name}
+            source_object["generation"] = source.generation
 
             preconditions = {}
             if (
-                if_generation_match is not None
-                and if_generation_match[index] is not None
+                if_source_generation_match is not None
+                and if_source_generation_match[index] is not None
             ):
-                preconditions["ifGenerationMatch"] = if_generation_match[index]
-
-            if (
-                if_metageneration_match is not None
-                and if_metageneration_match[index] is not None
-            ):
-                preconditions["ifMetagenerationMatch"] = if_metageneration_match[index]
+                preconditions["ifGenerationMatch"] = if_source_generation_match[index]
```

Comment:
This would then all simplify down to:

```python
            if source_generation is not None:
                source_object["objectPreconditions"] = {"ifGenerationMatch": source_generation}
```

Reply:
Done, thanks @tseaver! This is much clearer. I like it a lot.